File: solutions/javascript/robot-simulator/1/robot-simulator.js
/**
 * Class representing an error for invalid input.
 * @extends Error
 */
export class InvalidInputError extends Error {
  /**
   * Create an InvalidInputError.
   * @param {string} [message='Invalid Input'] - The error message.
   */
  constructor(message) {
    super();
    this.message = message || 'Invalid Input';
  }
}

/**
 * Class representing a Robot.
 */
export class Robot {
  /**
   * Get the instructions for the robot.
   * @param {string} s - The string of instructions.
   * @returns {Array<string>} The array of instruction methods.
   * @throws {InvalidInputError} If the instruction character is invalid.
   */
  static instructions(s) {
    return [...s].map((character) => {
      switch (character) {
        case 'L':
          return 'turnLeft';
        case 'R':
          return 'turnRight';
        case 'A':
          return 'advance';
        default:
          throw new InvalidInputError(
            `${character} is not a valid instruction character.`,
          );
      }
    });
  }

  /**
   * Create a Robot.
   */
  constructor() {
    this.coordinates = [0, 0];
    this.bearing = 'north';
  }

  /**
   * Set the direction of the robot.
   * @param {string} next - The next direction.
   * @throws {InvalidInputError} If the direction is invalid.
   */
  set direction(next) {
    const validDirections = ['north', 'south', 'east', 'west'];
    if (!validDirections.includes(next)) {
      throw new InvalidInputError('Invalid Robot Bearing');
    }

    this.bearing = next;
  }

  /**
   * Advance the robot in the current direction.
   */
  advance() {
    if (this.bearing === 'north') {
      this.coordinates[1] += 1;
    } else if (this.bearing === 'south') {
      this.coordinates[1] -= 1;
    } else if (this.bearing === 'east') {
      this.coordinates[0] += 1;
    } else if (this.bearing === 'west') {
      this.coordinates[0] -= 1;
    }
  }

  /**
   * Turn the robot to the left.
   */
  turnLeft() {
    if (this.bearing === 'north') {
      this.direction = 'west';
    } else if (this.bearing === 'south') {
      this.direction = 'east';
    } else if (this.bearing === 'east') {
      this.direction = 'north';
    } else if (this.bearing === 'west') {
      this.direction = 'south';
    }
  }

  /**
   * Turn the robot to the right.
   */
  turnRight() {
    if (this.bearing === 'north') {
      this.direction = 'east';
    } else if (this.bearing === 'south') {
      this.direction = 'west';
    } else if (this.bearing === 'east') {
      this.direction = 'south';
    } else if (this.bearing === 'west') {
      this.direction = 'north';
    }
  }

  /**
   * Place the robot at specific coordinates and direction.
   * @param {Object} args - The placement arguments.
   * @param {number} args.x - The x-coordinate.
   * @param {number} args.y - The y-coordinate.
   * @param {string} args.direction - The direction.
   */
  place(args) {
    this.coordinates = [args.x, args.y];
    this.direction = args.direction;
  }

  /**
   * Evaluate a string of instructions.
   * @param {string} s - The string of instructions.
   */
  evaluate(s) {
    Robot.instructions(s).forEach((instruction) => {
      this[instruction]();
    });
  }
}

File: solutions/elixir/robot-simulator/1/lib/robot_simulator.ex
defmodule RobotSimulator do
  @type robot() :: any()
  @type direction() :: :north | :east | :south | :west
  @type position() :: {integer(), integer()}
  defguard is_direction(direction) when direction in [:north, :east, :south, :west]
  defguard is_position(x, y) when is_integer(x) and is_integer(y)

  @doc """
  Create a Robot Simulator given an initial direction and position.

  Valid directions are: `:north`, `:east`, `:south`, `:west`
  """
  @spec create(direction, position) :: robot() | {:error, String.t()}
  def create(), do: %{direction: :north, position: {0, 0}}
  def create(direction, _) when not is_direction(direction), do: {:error, "invalid direction"}
  def create(direction, {x, y}) when is_position(x, y), do: %{direction: direction, position: {x, y}}
  def create(_, _) , do: {:error, "invalid position"}

  @doc """
  Simulate the robot's movement given a string of instructions.

  Valid instructions are: "R" (turn right), "L", (turn left), and "A" (advance)
  """
  @spec simulate(robot, instructions :: String.t()) :: robot() | {:error, String.t()}
  def simulate(robot, instructions) do
    String.graphemes(instructions)
    |> Enum.reduce(robot, &move/2)
  end

  def move("L", %{direction: :north} = robot), do: %{robot | direction: :west}
  def move("L", %{direction: :east} = robot), do: %{robot | direction: :north}
  def move("L", %{direction: :south} = robot), do: %{robot | direction: :east}
  def move("L", %{direction: :west} = robot), do: %{robot | direction: :south}

  def move("R", %{direction: :north} = robot), do: %{robot | direction: :east}
  def move("R", %{direction: :east} = robot), do: %{robot | direction: :south}
  def move("R", %{direction: :south} = robot), do: %{robot | direction: :west}
  def move("R", %{direction: :west} = robot), do: %{robot | direction: :north}

  def move("A", %{direction: :north, position: {x, y}} = robot), do: %{robot | position: {x, y + 1}}
  def move("A", %{direction: :east, position: {x, y}} = robot), do: %{robot | position: {x + 1, y}}
  def move("A", %{direction: :south, position: {x, y}} = robot), do: %{robot | position: {x, y - 1}}
  def move("A", %{direction: :west, position: {x, y}} = robot), do: %{robot | position: {x - 1, y}}
  def move(_, _), do: {:error, "invalid instruction" }

  @doc """
  Return the robot's direction.

  Valid directions are: `:north`, `:east`, `:south`, `:west`
  """
  @spec direction(robot) :: direction()
  def direction(robot) do
    robot.direction
  end

  @doc """
  Return the robot's position.
  """
  @spec position(robot) :: position()
  def position(robot) do
    robot.position
  end
end
